### PR TITLE
fix(developer): fix breakage from emscripten 3.1.44 🙀

### PR DESCRIPTION
**Pull request by srl295**
- use wasmExports.malloc if available otherwise asm.malloc

see:

https://github.com/emscripten-core/emscripten/blob/main/ChangeLog.md#3144---072523

Note, emscripten 3.1.43 and prior may give this warning:

> warning: invalid item in EXPORTED_RUNTIME_METHODS: wasmExports

@keymanapp-test-bot skip

### CODE REVIEW

File: developer/src/kmcmplib/src/meson.build [hunk 1]
```diff
@@ -28,7 +28,7 @@ if cpp_compiler.get_id() == 'emscripten'
   # wasm-exceptions supported in Node 18+, Chrome 95+, Firefox 100+, Safari 15.2+
   flags += ['-fwasm-exceptions']
   lib_links = ['--whole-archive', '-sMODULARIZE', '-sEXPORT_ES6']
-  links += ['-fwasm-exceptions', '--bind', '-sEXPORTED_RUNTIME_METHODS=[\'UTF8ToString\']']
+  links += ['-fwasm-exceptions', '--bind', '-sEXPORTED_RUNTIME_METHODS=[\'UTF8ToString\',\'wasmExports\']']
```
Comment (darcywong00) [inline developer/src/kmcmplib/src/meson.build:31]: Doesn't this ensure `this.Module.wasmExports` is always available? Do we still need to fallback to Module.asm?

Comment (srl295) [reply developer/src/kmcmplib/src/meson.build:31]: `wasmExports` is only available on 3.1.44.  Hooray for a smooth transition! 😆 

on 3.1.43 we get:

> warning: invalid item in EXPORTED_RUNTIME_METHODS: wasmExports

There may be another way to do this as `asm` is apparently internal, but I'm not sure what it is.

Comment (darcywong00) [reply developer/src/kmcmplib/src/meson.build:31]: ... if we update the prereq version of emcripten from 3.1.32 to 3.1.44

https://github.com/keymanapp/keyman/blob/master/docs/build/windows.md#web-dependencies

**Comment by mcdurdin** [comment]
Wow emscripten removing published APIs in a patch release?!

**Comment by srl295** [comment]
> Wow emscripten removing published APIs in a patch release?!

`asm` was supposedly internal, though I'm not sure how else it was supposed to be accessed.

File: developer/src/kmcmplib/src/meson.build (hunk 1 above)
Comment (srl295) [reply developer/src/kmcmplib/src/meson.build:31]: @darcywong00 as written here it works either way.

Comment (mcdurdin) [reply developer/src/kmcmplib/src/meson.build:31]: > There may be another way to do this as `asm` is apparently internal, but I'm not sure what it is.

We'd need to conditionalize it on the emscripten version which is nicely self-documenting in the meson.build. Something like:

```suggestion
  links += ['-fwasm-exceptions', '--bind']
  if cpp_compiler.version() >= '3.1.44'
    # emscripten 3.1.44 removes .asm object and so we need to export `wasmExports`
    # #9375; https://github.com/emscripten-core/emscripten/blob/main/ChangeLog.md#3144---072523
    links += ['-sEXPORTED_RUNTIME_METHODS=[\'UTF8ToString\',\'wasmExports\']']
  else
    # emscripten < 3.1.44 does not include `wasmExports`
    links += ['-sEXPORTED_RUNTIME_METHODS=[\'UTF8ToString\']']
  endif
```

Comment (srl295) [reply developer/src/kmcmplib/src/meson.build:31]: why? the wasmExports message is only a warning.

Comment (mcdurdin) [reply developer/src/kmcmplib/src/meson.build]: My goal is zero messages in our builds (it's a stretch goal for sure). Every extraneous warning is more noise to sift through for actual real warnings which need to be addressed.

**Review by mcdurdin — APPROVED** [review]

File: developer/src/kmc-kmn/src/compiler/compiler.ts [hunk 2]
```diff
@@ -237,7 +237,13 @@ export class KmnCompiler implements UnicodeSetParser {
       return null;
     }
 
-    const buf = this.Module.asm.malloc(rangeCount * 2 * this.Module.HEAPU32.BYTES_PER_ELEMENT);
+    let malloc = this.Module?.wasmExports?.malloc || this.Module?.asm?.malloc;
```
[the remark this answers is not in the document]
Comment (srl295) [reply developer/src/kmc-kmn/src/compiler/compiler.ts:240]: maks sense

File: developer/src/kmc-kmn/src/compiler/compiler.ts [hunk 3]
```diff
@@ -237,7 +237,13 @@ export class KmnCompiler implements UnicodeSetParser {
...
+
+    if (!malloc) {
+      throw new Error(`Internal Error: missing wasmExports.malloc() / asm.malloc()`);
+    }
```
[the remark this answers is not in the document]
Comment (srl295) [reply developer/src/kmc-kmn/src/compiler/compiler.ts:244]:
```suggestion
```

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 17.0.151-alpha](https://keyman.com/downloads/releases/alpha/17.0.151)

**Pull request merged**